Math packages: gf2x, ntl, flintlib, arb, eclib

tornaria opened the pull request:
<!-- Mark items with [x] where applicable -->

#### General
- [x] This is a new package and it conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements)

#### Have the results of the proposed changes been tested?
- [x] I use the packages affected by the proposed changes on a regular basis and confirm this PR works for me

#### Notes
- This includes: `gf2x` (see also #29783), `ntl`, `flintlib`, `arb`, `eclib`.
- Everything is tested and check pass in x86_64, x86_64-musl, i686.

EDIT: split in smaller independent PR to ease review.

abenson commented:
> This continues #29997.

Just add to that one.  At this point the two conflict -- it's easier to track when it's all in the same PR.

tornaria commented:
> > This continues #29997.
> 
> Just add to that one. At this point the two conflict -- it's easier to track when it's all in the same PR.

You are right. I rebased in several different and independent PR.

dkwo commented:
I did not test `eclib`, the rest looks good to me.

ericonr commented on srcpkgs/arb/template:
```diff
@@ -0,0 +1,29 @@
+pkgname=arb
+version=2.19.0
+revision=1
+build_style="configure"
```
```suggestion
build_style=configure
```

Style nit

ericonr commented on srcpkgs/arb/template:
```diff
@@ -0,0 +1,29 @@
+homepage="https://arblib.org"
+distfiles="https://github.com/fredrik-johansson/arb/archive/${version}.tar.gz"
+checksum=0aec6b492b6e9a543bdb3287a91f976951e2ba74fd4de942e692e21f7edbcf13
+nocross=yes
```
Do you have  a simple comment you can add here explaining why?

tornaria replied:
I don't remember from the top of my head; I did try it and something didn't work. I will retry it and comment.

tornaria replied:
- ntl has a custom configure perl script which runs binaries built for target.
- flintlib and arb should cross-compile fine once ntl is packaged

ericonr commented on srcpkgs/arb/template:
```diff
@@ -0,0 +1,29 @@
+checksum=0aec6b492b6e9a543bdb3287a91f976951e2ba74fd4de942e692e21f7edbcf13
+nocross=yes
+
+if [ "$XBPS_TARGET_LIBC" = "musl" ]; then
+	CFLAGS="-D_GNU_SOURCE"
+fi
```
Doesn't need to be conditional, setting this for glibc shouldn't hurt anything.

ericonr commented on srcpkgs/arb/template:
```diff
@@ -0,0 +1,29 @@
+
+arb-devel_package() {
+	depends="arb-${version}_${revision}"
+	depends+=" flintlib-devel"
```
```suggestion
	depends+=" ${makedepends}"
```

tornaria replied:
Note that many times `makedepends` has more packages than `depends` for the -devel package, just not in this case, that's why I tried to be explicit about it.

ericonr replied:
Ok, that's fair. Can leave as is.

ericonr commented on srcpkgs/arb/template:
```diff
@@ -0,0 +1,29 @@
+fi
+
+arb-devel_package() {
+	depends="arb-${version}_${revision}"
```
```suggestion
	depends="arb>=${version}_${revision}"
```

tornaria replied:
Somehow I feel uneasy about that.

ericonr replied:
This is done because otherwise parsing is ambiguous and that sucks.

ericonr replied:
And it helps with updates, in some occasions

ericonr commented on srcpkgs/arb/template:
```diff
@@ -0,0 +1,29 @@
+	depends+=" flintlib-devel"
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove "usr/include"
```
```suggestion
		vmove usr/include
```

Style nit

ericonr commented on srcpkgs/eclib/template:
```diff
@@ -0,0 +1,30 @@
+checksum=b742e322f5aeffef7256c05a9b9f744a8f34cb52777c62764cce07a49b1d32fa
+
+pre_configure() {
+	./autogen.sh
```
We usually try to use `autoreconf -fi`, does their script do more than that?

tornaria replied:
```
aclocal
libtoolize --automake
automake --add-missing
autoconf
```

I'll try to check if `autoreconf -fi` is enough.

tornaria replied:
`autoreconf -fi` is ok so I changed to that.

ericonr commented on srcpkgs/eclib/template:
```diff
@@ -0,0 +1,30 @@
+}
+
+eclib-devel_package() {
+	depends="eclib-${version}_${revision}"
```
```suggestion
	depends="eclib>=${version}_${revision}"
```

ericonr commented on srcpkgs/eclib/template:
```diff
@@ -0,0 +1,30 @@
+	depends+=" ntl-devel"
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove "usr/include"
+		vmove "usr/lib/*.a"
+		vmove "usr/lib/*.so"
+		vmove "usr/lib/pkgconfig"
```
Same style nit, remove quotes when not globbing.

ericonr commented on srcpkgs/flintlib/template:
```diff
@@ -0,0 +1,30 @@
+checksum=186e2fd9ab67df8a05b122fb018269b382e4babcdb17353c4be1fe364dca481e
+nocross=yes
+
+if [ "$XBPS_TARGET_LIBC" = "musl" ]; then
+	CFLAGS="-D_GNU_SOURCE"
+fi
```
Same as previous

ericonr commented on srcpkgs/flintlib/template:
```diff
@@ -0,0 +1,30 @@
+homepage="https://flintlib.org"
+distfiles="https://flintlib.org/flint-${version}.tar.gz"
+checksum=186e2fd9ab67df8a05b122fb018269b382e4babcdb17353c4be1fe364dca481e
+nocross=yes
```
Same as previous

ericonr commented on srcpkgs/flintlib/template:
```diff
@@ -0,0 +1,30 @@
+	depends+=" mpfr-devel"
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove "usr/include"
```
Same as previous

ericonr commented on srcpkgs/flintlib/template:
```diff
@@ -0,0 +1,30 @@
+fi
+
+flintlib-devel_package() {
+	depends="flintlib-${version}_${revision}"
```
```suggestion
	depends="flintlib>=${version}_${revision}"
```

ericonr commented on srcpkgs/flintlib/update:
```diff
@@ -0,0 +1,2 @@
+pkgname=flint
```
Maybe mention in commit that name was picked because of existing package named flint.

tornaria replied:
made a comment in the `update` file itself, hope that's ok

ericonr commented on srcpkgs/gf2x/template:
```diff
@@ -0,0 +1,34 @@
+version=1.3.0
+revision=1
+wrksrc=gf2x-gf2x-${version}
+build_style="gnu-configure"
```
```suggestion
build_style=gnu-configure
```

ericonr commented on srcpkgs/gf2x/template:
```diff
@@ -0,0 +1,34 @@
+
+build_options="native"
+
+if [ ! "$build_option_native" ]; then
```
```suggestion
if [ -z "$build_option_native" ]; then
```

ericonr commented on srcpkgs/gf2x/template:
```diff
@@ -0,0 +1,34 @@
+distfiles="https://gitlab.inria.fr/gf2x/gf2x/-/archive/gf2x-${version}/gf2x-gf2x-${version}.tar.gz"
+checksum=11bcf98b620c60c2ee3b4460b02b7be741f14cfdc26b542f22c92950926575e0
+
+build_options="native"
```
Please add a description for this option, it isn't obvious what it does. `desc_option_native="..."`

I would name it sse3 directly, maybe. Then you could do `configure_args="$(vopt_enable sse3)"`.

Even if you keep the native name, you can still do `configure_args="$(vopt_if native '' '--disable-sse3')"` and have configure args together with build style instead of down here.

tornaria replied:
I added this to `common/options.description` in #29997 (see 2274429). There is a discussion about this in that PR.

My intention is to have an easy way to recompile math software for native arch. Depending on the package sometimes is sse3, sometimes is more. In this case the library also uses sse4.1 and pclmul, but they are dependents so `--disable-sse3` is enough to disable all of that is not generic (x86_64 implies sse2 but nothing more).

Not disabling sse3 as above means that the library will be compiled with -msse3, -msse41, -mpclmul, ... and maybe also -march=native, etc, so that all the options available in the current cpu will be used.

In general, different software have different features, oftentimes they have custom written asm for different architectures so selecting build option "native" will not necessarily be equivalent to just using `-march=native`.

In many cases (like here) one has to be careful because native is the default and it is necessary to make sure the binary actually is generic enough for x86_64 (I'm not sure how to test this -- can a VM disable some sse levels even if the host has them? I have an old core 2 quad (q9550) working (even with 8GB RAM) but that already has sse3!

ericonr replied:
AFAIK you can test, but I'm not sure it's necessary. We can wait for bug reports, IMO.

 Please import the commit into this PR, then :)

ericonr replied:
I wonder if `native_build` would be better.

ericonr commented on srcpkgs/gf2x/template:
```diff
@@ -0,0 +1,34 @@
+}
+
+gf2x-devel_package() {
+	depends="gf2x-${version}_${revision}"
```
```suggestion
	depends="gf2x>=${version}_${revision}"
```

ericonr replied:
I remembered it just now, but the default style is actually `${sourcepkg}>=...`

tornaria replied:
done for all 5 templates

ericonr commented on srcpkgs/gf2x/template:
```diff
@@ -0,0 +1,34 @@
+	depends="gf2x-${version}_${revision}"
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove "usr/include"
+		vmove "usr/lib/*.a"
+		vmove "usr/lib/*.so"
+		vmove "usr/lib/pkgconfig"
```
Same as above

ericonr commented on srcpkgs/ntl/template:
```diff
@@ -0,0 +1,36 @@
+version=11.4.4
+revision=1
+build_wrksrc="src"
+build_style="configure"
```
Same as above

ericonr commented on srcpkgs/ntl/template:
```diff
@@ -0,0 +1,36 @@
+homepage="https://libntl.org"
+distfiles="https://libntl.org/ntl-${version}.tar.gz"
+checksum=9d7f6e82e11a409f151c0de2deb08c0d763baf9834fddfd432bf3d218f8021db
+nocross=yes
```
Same as above

ericonr commented on srcpkgs/ntl/template:
```diff
@@ -0,0 +1,36 @@
+checksum=9d7f6e82e11a409f151c0de2deb08c0d763baf9834fddfd432bf3d218f8021db
+nocross=yes
+
+build_options="native"
```
Same as above

ericonr commented on srcpkgs/ntl/template:
```diff
@@ -0,0 +1,36 @@
+}
+
+ntl-devel_package() {
+	depends="ntl-${version}_${revision}"
```
Same as above

ericonr commented on srcpkgs/ntl/template:
```diff
@@ -0,0 +1,36 @@
+	depends+=" gmp-devel gf2x-devel"
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove "usr/include"
+		vmove "usr/lib/*.a"
+		vmove "usr/lib/*.so"
+		vmove "usr/share"
```
Same as above.

ericonr commented on srcpkgs/ntl/update:
```diff
@@ -0,0 +1 @@
+site="${homepage}/download.html"
```
Maybe comment in template that homepage is used in update file too? So it isn't clobbered accidentally.

tornaria replied:
Isn't `"${homepage}"` the default for `site` in any case?

ericonr replied:
Right, but if homepage moves to a github website or whatever the path won't exist anymore.

tornaria replied:
I changed site to be explicit and not use `${homepage}`, that's what seems to be done in most templates. I didn't see any notice in the template about syncing with `update`, but I think if the homepage changes for some reason and update-check can't find site anymore it will complain about "no version found" or something like that.

ericonr reviewed (commented):
Remember to squash all the commits for eclib :)

tornaria commented:
General comment for @ericonr : Thanks for your detailed review. I will apply all suggestions you made, and update to newer versions for arb, eclib, flintlib, ntl.

I hope this can be merged into void -- I want to help include more math software in void, I can maintain, update, etc.

ericonr commented:
My `_GNU_SOURCE` suggestion was apparently a bad idea for one of the packages, feel free to move it back to a conditional for musl.

Turns out glibc will define its own `fdiv` function if you include `<tgmath.h>` with `_GNU_SOURCE` defined, so we can't define it for glibc to avoid polluting the namespace and causing conflicts.

tornaria commented:
> My `_GNU_SOURCE` suggestion was apparently a bad idea for one of the packages, feel free to move it back to a conditional for musl.

I actually went ahead and fixed it at the source: https://github.com/wbhart/flint2/pull/988 and https://github.com/fredrik-johansson/arb/pull/376. My new version will include those as patches which they will hopefully be merged upstream.

tornaria commented:
Sorry for the double force-push. I realized a minute too late that I was missing a requested change.

I believe I have addressed all the suggestions.

tornaria commented:
The "fix" for the `_GNU_SOURCE` issue was bad, which caused the musl build to fail check.

Based on this and the comments in https://github.com/wbhart/flint2/pull/988, I removed all patches and reverted back to using `CFLAGS=-D_GNU_SOURCE` in the template, with the care that for `arb` it has to be conditional to avoid a conflict.

Now all tests should pass... :crossed_fingers:

ericonr commented on srcpkgs/gf2x/template:
```diff
@@ -0,0 +1,35 @@
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove usr/include
+		vmove usr/lib/*.a
+		vmove usr/lib/*.so
```
`vmove "usr/lib/*.a"`

ericonr commented on srcpkgs/ntl/template:
```diff
@@ -0,0 +1,36 @@
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove usr/include
+		vmove usr/lib/*.a
+		vmove usr/lib/*.so
```
same

ericonr commented on srcpkgs/flintlib/template:
```diff
@@ -0,0 +1,27 @@
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove usr/include
+		vmove usr/lib/*.a
+		vmove usr/lib/*.so
```
Same

ericonr commented on srcpkgs/arb/template:
```diff
@@ -0,0 +1,29 @@
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove usr/include
+		vmove usr/lib/*.a
+		vmove usr/lib/*.so
```
Same

ericonr commented on srcpkgs/flintlib/template:
```diff
@@ -0,0 +1,27 @@
+CFLAGS="-D_GNU_SOURCE" # needed for cpu_set_t
+
+flintlib-devel_package() {
+	depends="${sourcepkg}>=${version}_${revision}"
+	depends+=" mpfr-devel"
```
Make this  a single line

ericonr commented on srcpkgs/arb/template:
```diff
@@ -0,0 +1,29 @@
+fi
+
+arb-devel_package() {
+	depends="${sourcepkg}>=${version}_${revision}"
+	depends+=" ${makedepends}"
```
Same here

ericonr commented on srcpkgs/eclib/template:
```diff
@@ -0,0 +1,30 @@
+}
+
+eclib-devel_package() {
+	depends="${sourcepkg}>=${version}_${revision}"
+	depends+=" ntl-devel"
+	short_desc+=" - development files"
+	pkg_install() {
+		vmove usr/include
+		vmove usr/lib/*.a
+		vmove usr/lib/*.so
+		vmove usr/lib/pkgconfig
+	}
```
Same here

tornaria commented:
All done.

ericonr reviewed (approved):
LGTM, going to leave open for a bit so anyone can leave thoughts before merging.

tornaria commented:
@dkwo: do you want to test this? I've successfully built sagemath-9.4.rc1 with new pari + new giac + this PR.

Note that you have to rebase this PR before compiling so it gets the pari update (otherwise it will complain about a missing soname and refuse to pkg).

dkwo commented:
@tornaria Thanks, I noticed that.
Didn't you need `configure_args+=" --with-readline --with-gmp"` in pari to use with sage?
I will rebuild sage as well soon, but I think you can go ahead with these PR's in the meantime.

tornaria commented:
> @tornaria Thanks, I noticed that.
> Didn't you need `configure_args+=" --with-readline --with-gmp"` in pari to use with sage?

Not really, they are default on when `readline-devel` and `gmp-devel` are installed.

If you start our new gp (already in the void repo) you can see in the banner:
```
                                                                 GP/PARI CALCULATOR Version 2.13.2 (released)
                                                         amd64 running linux (x86-64/GMP-6.2.1 kernel) 64-bit version
                                                           compiled: Aug 15 2021, gcc version 10.2.1 20201203 (GCC)
                                                                          threading engine: pthread
                                                                (readline v8.1 enabled, extended help enabled)
```
Here gmp, readline are enabled. Also pthreads, which is required in sage starting in 9.4.

dkwo commented:
I see, you are right.

dkwo commented:
@tornaria Btw, I've also just built `sage-9.4.rc2` with this, the other four math package and the giac PR's :)

ericonr commented:
Ping me for the next math package PR we should review ;)
